clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/model/UltrablueProtocol.kt

```diff
@@ -30,36 +60,166 @@ const val CREDENTIAL_ACTIVATION_ASSERT = 6
     update the state machine, and resume the protocol.
  */
 
-class UltrablueProtocol(private val activity: MainActivity, private var device: Device, private val readMsg: (String) -> Unit, private var writeMsg: (String, ByteArray) -> Unit) {
-    private var state: ProtocolStep = REGISTRATION_READ
+class UltrablueProtocol(
+    private val activity: MainActivity,
+    private var enroll: Boolean = false,
+    private var device: Device,
+    private val logger: Logger?,
+    private val readMsg: (String) -> Unit,
+    private var writeMsg: (String, ByteArray) -> Unit,
+    private var onCompletion: (Boolean) -> Unit
+) {
+    private var state: ProtocolStep = if (enroll) { EK_READ } else { AUTHENTICATION_READ }
     private var message = byteArrayOf()
 
-    init {
+    private val rand = SecureRandom()
+    private var ek = RegistrationDataModel(device.ekn, device.eke.toUInt(), device.ekcert, device.secret.isNotEmpty()) // If enrolling a device, this field is uninitialized, but will be after EK_READ.
+    private var credentialActivationSecret: ByteArray? = null
+    private var encodedAttestationKey: ByteArray? = null
+    private var encodedPlatformParameters: ByteArray? = null
+    private val attestationNonce = ByteArray(16)
+    private var encodedPCRs: ByteArray? = null
+    private var attestationResponse: AttestationResponse? = null
+
+
+    fun start() {
         resume()
     }
 
     @OptIn(ExperimentalSerializationApi::class)
     private fun resume() {
         when (state) {
-            REGISTRATION_READ -> readMsg("EkPub and EkCert")
-            REGISTRATION -> {
-                val key = Cbor.decodeFromByteArray<EKModel>(message)
-                device.name = "device" + device.uid
-                device.ekn = key.N
-                device.eke = key.E.toInt()
-                device.ekcert = key.Cert
-                activity.viewModel.insert(device)
-                writeMsg("registration confirmation", byteArrayOf(-10)) // Registration success, CBOR encoded
+            EK_READ -> readMsg(activity.getString(R.string.ek_pub_cert))
+            EK_DECODE -> {
+                ek = Cbor.decodeFromByteArray(message)
+                state += 1
+                resume()
             }
-            AUTHENTICATION_READ -> readMsg("authentication nonce")
+            AUTHENTICATION_READ -> readMsg(activity.getString(R.string.auth_nonce))
             AUTHENTICATION -> {
-                // When encryption will be implemented, we'll first need to decrypt the nonce here.
-                writeMsg("Nonce", message)
+                if (message.size != 24) {
+                    logger?.push(CLog("Invalid nonce length. Make sure you ran the attestation server without the --enroll flag.", false))
```

Comment:
I can do that by wrapping the CBOR decoding into a try/catch block, so that it will throw if a wrong data structure is sent, but I'll still have to make the detection here as there's only one Bluetooth characteristic, thus the server 'chooses' what to send, and the only thing we can do is check that it matches what we expect to receive.
It happens that if the wrong mode is choosen, we will figure it at this point.

Do you prefer the try/catch way?

Reply:
I'd prefer the later as it will catch any kind of protocol misuse and not only the invalid nonce one.

Reply:
Agreed, catch all errors, and still try to provide a helpful error message ("error blabla. Check that the server and client are both running in the same mode.").

Another option is to redesign the message format to prefix them with a message type, and check that we get message of the expected type at each step. The technique is known as TLV and is a good design approach to evolve your protocol later: https://en.wikipedia.org/wiki/Type%E2%80%93length%E2%80%93value

Let's not redesign at this stage, but feel free to create an issue to track this enhancement.